### PR TITLE
[Feature] [ShenYu]<Apache ShenYu (incubating) plugin support dubbo proxy>(#8869)

### PR DESCRIPTION
**Pull request by hutaishi**
Add skywaliking plugin for support Apache ShenYu (incubating) gateway.

- [x] Add a test case for the new plugin, refer to [the doc](https://github.com/apache/skywalking-java/blob/main/docs/en/setup/service-agent/java-agent/Plugin-test.md)
- [x] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.
- [x] If it's UI related, attach the screenshots below.

### topology
image

### trace
image

### CODE REVIEW

**Comment by wu-sheng** [comment]
2 questions
1. Which parts are pure Dubbo plugin enhancement? Tbgis part shoud be in another PR
2. Why you need a snapshot continue inside Dubbo plugin?You could use a local span to continue in Shenyu. Binding 2 plugins with a context key is not recommended, only when has to be done, I feel this is not the case.

**Comment by hutaishi** [comment]
> 2 questions
> 
> 1. Which parts are pure Dubbo plugin enhancement? Tbgis part shoud be in another PR
> 2. Why you need a snapshot continue inside Dubbo plugin?You could use a local span to continue in Shenyu. Binding 2 plugins with a context key is not recommended, only when has to be done, I feel this is not the case.

Hello, I'd like to ask the second question. I tried to use localspan, but due to cross threading, abstracttracercontext is empty when closing localspan.

image




image

**Comment by wu-sheng** [comment]
I  not suggesting local span in Dubbo plugin, I suggest you to add a new local span in your own codes and in the same thread of Dubbo filter. Then they are linked automatically.

**Comment by hutaishi** [comment]
> I not suggesting local span in Dubbo plugin, I suggest you to add a new local span in your own codes and in the same thread of Dubbo filter. Then they are linked automatically.


 I understand in an instant. Thank you for your guidance.
successful.
image

I'll change it in your way later.

**Comment by hutaishi** [comment]
> 2 questions
> 
> 1. Which parts are pure Dubbo plugin enhancement? Tbgis part shoud be in another PR
> 2. Why you need a snapshot continue inside Dubbo plugin?You could use a local span to continue in Shenyu. Binding 2 plugins with a context key is not recommended, only when has to be done, I feel this is not the case.

now:
Only the Shenyu plugin is modified and a localspan is created to concat Dubbo

File: apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/ApacheDubboPluginInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.lang.reflect.Method;
+
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.plugin.shenyu.v24x.util.CommonUtil;
+
+/**
+ * shenyu AlibabaDubboPlugin Interceptor.
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/ApacheDubboPluginInterceptor.java:29]: Dubbo doesn't belong to Alibaba for a long time. Please fix relative comments.

File: apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+public class AlibabaDubboPluginInstrumentation extends ClassInstanceMethodsEnhancePluginDefine {
+    @Override
+    protected ClassMatch enhanceClass() {
+        return byName("org.apache.shenyu.plugin.alibaba.dubbo.AlibabaDubboPlugin");
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java:34]: Why does Shenyu require old Dubbo? As a proxy, this should be transparent, isn't it?

**Comment by wu-sheng** [comment]
Besides Alibaba legacy Dubbo question, the plugin test should be added too.

File: apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java (hunk 2 above)
Comment (hutaishi) [reply apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java:34]: I have tested that Dubbo 2.7x client can access the old version of Dubbo, but Shenyu is based on two plug-ins, not one.
I'll give a suggestion to Shenyu community.

**Comment by hutaishi** [comment]
> Besides Alibaba legacy Dubbo question, the plugin test should be added too.

thinks, I will add test.

File: apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java (hunk 2 above)
Comment (wu-sheng) [reply apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java:34]: Alibaba Dubbo is like 3 years old, I guess? And considering no one is maintaining it, and fewer people are going to use it, I would suggest we remove that support.
Even for SkyWalking's native Alibaba Dubbo plugin, once we received a report saying there is a bug, the official solution would be to remove it too.

Comment (hutaishi) [reply apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java:34]: > Alibaba Dubbo is like 3 years old, I guess? And considering no one is maintaining it, and fewer people are going to use it, I would suggest we remove that support. Even for SkyWalking's native Alibaba Dubbo plugin, once we received a report saying there is a bug, the official solution would be to remove it too.

Agree. In addition, the new version of Dubbo client can also call the old duboo service, although I haven't fully tested it

I will remove support for the old version.

Comment (wu-sheng) [reply apm-sniffer/optional-plugins/shenyu-2.4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/shenyu/v24x/define/AlibabaDubboPluginInstrumentation.java:34]: > Agree. In addition, the new version of Dubbo client can also call the old duboo service, although I haven't fully tested it

I think the protocol is not unchanged, RPC frameworks are keeping forward compatibility seriously, I assume Dubbo community should consider this already. 
If you or Shenyu community has concerns, you could submit a question to Dubbo to confirm, I think they would be happy to confirm and see 3 projects closing integration.

**Comment by hutaishi** [comment]
> Besides Alibaba legacy Dubbo question, the plugin test should be added too.

Thanks to the mentor's guidance, it has been modified and passed the local test, the UI works fine，Please help review again.

File: test/plugin/scenarios/shenyu-2.4.x-dubbo-scenario/config/expectedData.yaml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+segmentItems:
+  - serviceName: shenyu-gateway
+    segmentSize: nq 0
+    segments:
+      - segmentId: not null
+        spans:
+          - operationName: test.apache.skywalking.apm.testcase.shenyu.http.services.OrderService.$invoke(String,String[],Object[])
```
Comment (wu-sheng) [inline test/plugin/scenarios/shenyu-2.4.x-dubbo-scenario/config/expectedData.yaml:22]: Is this `OrderService.$invoke` expected? Is this an auto-generated method?

Comment (hutaishi) [reply test/plugin/scenarios/shenyu-2.4.x-dubbo-scenario/config/expectedData.yaml:22]: Yes， skywalkinig dubbo plugin aotu-generated.    Generic service in dubbo.

**Comment by hutaishi** [comment]
> General good, once you confirmed whether the operation names fit your expectation.


This operation name is generated by the skywalking Dubbo plugin, which does not support generalization calls well.

**Comment by wu-sheng** [comment]
> > General good, once you confirmed whether the operation names fit your expectation.
> 
> This operation name is generated by the skywalking Dubbo plugin, which does not support generalization calls well.

A tip, our doc doesn't mention HTTP/gRPC or some other protocols(you supported) are/is supported or not, you may be better to document this on your side. @hutaishi

**Comment by hutaishi** [comment]
> > > General good, once you confirmed whether the operation names fit your expectation.
> > 
> > 
> > This operation name is generated by the skywalking Dubbo plugin, which does not support generalization calls well.
> 
> A tip, our doc doesn't mention HTTP/gRPC or some other protocols(you supported) are/is supported or not, you may be better to document this on your side. @hutaishi

Thank you for your guidance and tip, in fact, I would like to propose an issue to support other rpc and improve the documentation, you see whether it is feasible

**Comment by wu-sheng** [comment]
OK, generally about the details, Shenyu is better than us to host, because you would keep your eyes on the changes with time. Here (in SkyWalking), it will be easy to outdated, eventually, it will harm your end users.

**Pull request merged**